Dev#01 : Common PR for all latest changes ( Aptos, GAuth, Sign Message, Onboarding, whatsapp/email OTP)

SrijanCoinDCX opened the pull request.

akathecoder commented on packages/core-js/src/authentication/email.ts:
```diff
@@ -0,0 +1,155 @@
+  ): Promise<
+    EmailSendOtpRequest | EmailResendOtpRequest | EmailVerifyOtpRequest
+  > {
+    const data: any = {};
```
@karansinglacoindcx please define a proper type.

akathecoder commented on packages/core-js/src/authentication/email.ts:
```diff
@@ -0,0 +1,155 @@
+      data,
+      client_signature: clientSignature,
+      type: 'ethsign',
+    } as any;
```
proper type missing

akathecoder commented on packages/core-js/src/authentication/whatsapp.ts:
```diff
@@ -0,0 +1,164 @@
+  ): Promise<
+    WhatsAppSendOtpRequest | WhatsAppResendOtpRequest | WhatsAppVerifyOtpRequest
+  > {
+    const data: any = {};
```
proper type

akathecoder commented on packages/core-js/src/authentication/social.ts:
```diff
@@ -0,0 +1,56 @@
+      redirect_uri: 'https://onboarding.oktostage.com/__/auth/handler',
+      response_type: 'id_token',
+      client_id:
+        '54780876714-t59u4t7r1pekdj3p54grd9nh4rfg8qvd.apps.googleusercontent.com',
```
put the client id in a constant dict in this class

akathecoder commented on packages/core-js/src/authentication/social.ts:
```diff
@@ -0,0 +1,56 @@
+      response_type: 'id_token',
+      client_id:
+        '54780876714-t59u4t7r1pekdj3p54grd9nh4rfg8qvd.apps.googleusercontent.com',
+      nonce: 'b703d535-bc46-4911-8aa3-25fb6c19e2ce',
```
what is this nonce. if this is supposed to be random, please use `uuidv4()`

akathecoder replied:
@SrijanCoinDCX

akathecoder commented on packages/core-js/src/core/oktoClientInputValidator.ts:
```diff
@@ -72,3 +88,22 @@ export const validateSessionConfig = (data: unknown) =>
+      return validatePhoneNumber(contact);
+    }
+    return null;
+  } catch (error) {
```
TS: 'error' is defined but never used.

akathecoder commented on packages/core-js/src/types/gateway/authenticate.ts:
```diff
@@ -6,6 +6,10 @@ export type AuthData =
   | {
       authToken: string;
       provider: 'okto';
+    }
+  | {
+      idToken: string;
+      provider: 'okto';
     };
```
@karansinglacoindcx what is the difference b/w 2nd and 3rd ?

akathecoder replied:
@karansinglacoindcx remove {
      authToken: string;
      provider: 'okto';
    } is not needed

akathecoder commented on packages/core-js/package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@okto_web3/core-js-sdk",
-  "version": "0.6.0",
+  "version": "0.6.1-dev.2",
```
please remove dev tags from package.json
@karansinglacoindcx @SrijanCoinDCX

akathecoder commented on packages/react/src/core/index.ts:
```diff
@@ -25,6 +39,19 @@ class OktoClient extends OktoCoreClient {
+    this.authWebView = new OktoAuthWebView(this.webViewManager, authHandler);
+  }
+
+  public authenticateWithWebView(options: WebViewOptions = {}): Promise<any> {
```
please don't use any as return type

akathecoder commented on packages/react/src/hooks/useOkto.ts:
```diff
@@ -10,3 +10,8 @@ export function useOkto(): OktoClient {
   return client;
 }
+
+export function useAuthWebView() {
+  const client = useOkto();
+  return client['authWebView']; // Access the private `authWebView` instance
+}
```
@SrijanCoinDCX why two functions with similar names ?

`useWebViewAuth()`  and `useAuthWebView()`

akathecoder replied:
@SrijanCoinDCX remove this

akathecoder commented on packages/react/src/webview/auth/authRequestHandler.ts:
```diff
@@ -0,0 +1,297 @@
+
+      case 'verify_otp':
+        console.log('Verifying OTP:', actualData.data.otp);
+        const response = await this.handleVerifyOtp(
+          actualData.data.provider == 'email'
+            ? actualData.data.email
+            : actualData.data.whatsapp_number,
+          actualData.data.otp,
+          actualData.data.provider,
+          actualData.data.token,
+          baseResponse,
+        );
```
eslint: Unexpected lexical declaration in case block.

akathecoder commented on packages/react/src/webview/auth/authRequestHandler.ts:
```diff
@@ -0,0 +1,297 @@
+        throw new Error('OktoClient is not initialized');
+      }
+      const response = await this.oktoClient.sendOTP(contact, provider);
+      var payload =
+        provider == 'email'
+          ? { provider: provider, email: contact, token: response?.token }
+          : {
+              provider: provider,
+              whatsapp_number: contact,
+              token: response?.token,
+            };
```
eslint: Unexpected var, use let or const instead.

akathecoder commented on packages/react/src/webview/auth/authWebView.ts:
```diff
@@ -0,0 +1,103 @@
+        url = `${DEFAULT_WEBVIEW_URL}?app=OKTO_WEB&origin=${window.location.origin}`,
+        width = 500,
+        height = 800,
+        onSuccess,
```
TS: 'onSuccess' is declared but its value is never read.

akathecoder commented on packages/react/src/webview/auth/authWebView.ts:
```diff
@@ -0,0 +1,103 @@
+    this.authRequestHandler = authRequestHandler;
+  }
+
+  public open(options: WebViewOptions = {}): Promise<any> {
```
any as return type

akathecoder commented on packages/react/src/webview/webViewManager.ts:
```diff
@@ -0,0 +1,444 @@
+  public sendRequest(method: string, data: any): Promise<any> {
+    return new Promise((resolve, reject) => {
+      if (!this.isWebViewOpen()) {
+        this.debug &&
+          console.warn(
+            '[WebViewManager] WebView not available for sending request',
+          );
```
eslint: Expected an assignment or function call and instead saw an expression.

akathecoder replied:
@SrijanCoinDCX it is in multiple lines of code across codebase. please fix all eslint errors

akathecoder replied:
++ @karansinglacoindcx similar errors in RN as well

akathecoder commented on packages/react-native/src/hooks/useOktoWebView.ts:
```diff
@@ -0,0 +1,26 @@
+ *
+ * @returns A function to open web view with the provided URL
+ */
+export function useOktoWebView() {
```
The hooks are different in both React & RN.
@karansinglacoindcx @SrijanCoinDCX please try to make them same, if possible

akathecoder replied:
@SrijanCoinDCX to update in react

akathecoder commented on packages/react-native/src/webview/authentication/authWebViewHandlers.ts:
```diff
@@ -0,0 +1,476 @@
+
+      switch (provider) {
+        case 'whatsapp':
+          const { whatsapp_number } = request.data;
```
eslint: Unexpected lexical declaration in case block.

akathecoder commented on packages/react-native/src/webview/authentication/authWebViewScreen.tsx:
```diff
@@ -0,0 +1,124 @@
+import type { WebViewParamList } from '../types.js';
+import { OktoClient } from '@okto_web3/core-js-sdk';
+import { AuthWebViewRequestHandler } from './authWebViewHandlers.js';
+import { clearStorage, getStorage } from '../../utils/storageUtils.js';
```
TS: All imports in import declaration are unused.

akathecoder commented on packages/react-native/package.json:
```diff
@@ -34,9 +34,21 @@
     "@noble/ciphers": "^1.2.1",
     "@okto_web3/core-js-sdk": "workspace:*",
+    "@react-navigation/native": "^7.1.6",
+    "@react-navigation/native-stack": "^7.3.10",
+    "@react-navigation/stack": "^7.2.10",
+    "@types/react-native": "^0.73.0",
     "@types/react-native-get-random-values": "^1.8.2",
+    "@types/react-navigation": "^3.4.0",
+    "@types/uuid": "^10.0.0",
+    "expo-clipboard": "^7.0.1",
+    "expo-web-browser": "^14.0.2",
+    "react-native": "^0.79.0",
     "react-native-get-random-values": "^1.11.0",
-    "react-native-mmkv": "^3.2.0"
+    "react-native-mmkv": "^3.2.0",
+    "react-native-safe-area-context": "^5.3.0",
+    "react-native-webview": "^13.6.4",
```
@karansinglacoindcx please make RN and expo related dependencies as peerDependencies

akathecoder commented on packages/react/package.json:
```diff
@@ -57,7 +57,8 @@
     "tsup": "^8.3.6",
-    "typescript": "^5.7.2"
+    "typescript": "^5.7.2",
+    "uuid": "^11.0.4"
```
@SrijanCoinDCX why is UUID a devDependencies?

akathecoder commented on package.json:
```diff
@@ -18,7 +18,8 @@
   "dependencies": {
-    "@changesets/cli": "^2.27.11"
+    "@changesets/cli": "^2.27.11",
+    "@okto_web3/react-sdk": "^0.4.1-dev.13"
```
why is a package a dependency in monorepo. please remove it

SrijanCoinDCX replied:
This was done automatically on linking the the packages locally removed.

akathecoder commented on packages/react/src/webview/auth/authWebView.ts:
```diff
@@ -0,0 +1,102 @@
+    this.authRequestHandler = authRequestHandler;
+  }
+
+  public open(options: WebViewOptions = {}): Promise<unknown> {
```
@SrijanCoinDCX make it void

akathecoder commented:
@karansinglacoindcx react-native dependencies

SrijanCoinDCX commented:
This pull request introduces new features and enhancements to the `@okto_web3` SDKs, focusing on blockchain transactions, NFT functionalities, and authentication mechanisms. It also includes minor updates to dependencies and package configurations.

### New Features and Enhancements:

#### Blockchain Transactions and NFT Support:
* Added support for executing Aptos raw transactions via the new `aptosRawTransaction` function in `@okto_web3/core-js-sdk`. This enables users to interact with the Aptos blockchain for raw transaction execution. (`packages/core-js/src/abstract/aptosRawTransaction.ts`, [packages/core-js/src/abstract/aptosRawTransaction.tsR1-R24](diffhunk://#diff-e363af3d33938d9c85bd0043839e0c2333ebd567af6d446924e17802703c424eR1-R24))
* Introduced `nftMint` and `nftCreateCollection` functions, allowing users to mint NFTs and create NFT collections programmatically. (`packages/core-js/src/abstract/nftMint.ts`, [[1]](diffhunk://#diff-6b05b77c8a3bdf57d6d72fc960b7ce968935b339e2c41a0446f5568ea8649b40R1-R24); `packages/core-js/src/abstract/nftCollectionCreation.ts`, [[2]](diffhunk://#diff-2df2093f1e11cb9b18d14d83b158ad187f2dbec9a97aa11eacc12e88bd596e21R1-R24)
* Renamed the `rawTransaction.ts` file to `evmRawTransaction.ts` and updated its function to handle EVM-specific raw transactions. (`packages/core-js/src/abstract/evmRawTransaction.ts`, [packages/core-js/src/abstract/evmRawTransaction.tsL2-R11](diffhunk://#diff-252f6896c7ecd4c553484d43239bf51e5882371f9c2cd659326c86e73e38e291L2-R11))

#### Authentication Enhancements:
* Added WhatsApp and Email OTP APIs to the core SDK, enabling multi-channel authentication. (`packages/core-js/src/api/auth.ts`, [packages/core-js/src/api/auth.tsR1-R151](diffhunk://#diff-4338b61eb2e3037f9e0fcf633cc0145b4a28cab6c502e58d131fab1a96fc9dd8R1-R151))
* Implemented `EmailAuthentication` and `WhatsAppAuthentication` classes for handling OTP-based authentication workflows. (`packages/core-js/src/authentication/email.ts`, [[1]](diffhunk://#diff-80b5f4befd1822e1e459ca18871b5c18351f13efbd7c57f6cf24dc4a39b34819R1-R155); `packages/core-js/src/authentication/index.ts`, [[2]](diffhunk://#diff-d084d93995ff77d5142ca427de7b93fd74ba0d633a9a6e3124e85f7b0a7eb282R1-R2)
* Introduced a `SocialAuthUrlGenerator` class to generate authentication URLs for social login providers like Google. (`packages/core-js/src/authentication/social.ts`, [packages/core-js/src/authentication/social.tsR1-R57](diffhunk://#diff-42f0127c4d7225127ff650eeb36b33824e261428cfd29b0d16943f0e165d00b2R1-R57))

### Configuration and Dependency Updates:
* Updated `package.json` to include type resolutions for React and React Native dependencies, ensuring compatibility. (`package.json`, [package.jsonR42-R46](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R42-R46))
* Changed the versioning in `@okto_web3/core-js-sdk` to a development version (`0.6.2-dev.3`). (`packages/core-js/package.json`, [packages/core-js/package.jsonL3-R3](diffhunk://#diff-feed3dc4c8abc58ceafa16570b0f7ea2a5e1b157f47e5dd2e9e18f9fdb160afeL3-R3))

These changes collectively enhance the SDK's functionality, making it more versatile for blockchain interactions, NFT management, and authentication workflows.

akathecoder reviewed (approved):
approved
